spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/security/IcebergHiveConnectorDelegationTokenProvider.java

```diff
@@ -0,0 +1,283 @@
+      Arrays.stream(sparkConf.getAllWithPrefix(targetPrefix))
+          .forEach(
+              entry -> {
+                if (entry._1.contains(URI_KEY)) {
+                  hiveConf.set(HiveConf.ConfVars.METASTOREURIS.varname, entry._2);
+                } else {
+                  hiveConf.set(entry._1, entry._2);
+                }
+              });
+      // The `RetryingHiveMetaStoreClient` may block the subsequent token obtaining,
+      // and `obtainDelegationTokens` is scheduled frequently, it's fine to disable
+      // the Hive retry mechanism.
+      hiveConf.set(HiveConf.ConfVars.METASTORE_FASTPATH.varname, "false");
+
+      return Optional.of(hiveConf);
+    } catch (Exception e) {
+      LOG.warn("Fail to create Hive Configuration for catalog {}: {}", catalogName, e.getMessage());
+      return Optional.empty();
+    }
+  }
+
+  /**
+   * Extracts the names of Iceberg catalogs from the provided Spark configuration. This method
+   * filters the Spark configuration entries to identify those that correspond to Iceberg catalogs.
+   * It checks if the catalog type matches either `SparkSessionCatalog` or `SparkCatalog` and
+   * collects their names.
+   *
+   * @param sparkConf The SparkConf object containing Spark configurations.
+   * @return A Set of Strings representing the names of Iceberg catalogs.
+   */
+  private Set<String> extractIcebergCatalogNames(SparkConf sparkConf) {
+    return Arrays.stream(sparkConf.getAllWithPrefix(SPARK_SQL_CATALOG_PREFIX))
+        .filter(
+            entry -> {
+              String val = entry._2();
+              return val.contains(SparkSessionCatalog.class.getName())
+                  || val.contains(SparkCatalog.class.getName());
+            })
+        .map(Tuple2::_1)
+        .collect(Collectors.toSet());
+  }
+
+  @Override
+  public boolean delegationTokensRequired(SparkConf sparkConf, Configuration hadoopConf) {
+    return !getRequireTokenCatalogs(sparkConf, hadoopConf).isEmpty();
+  }
+
+  private Set<String> getRequireTokenCatalogs(SparkConf sparkConf, Configuration hadoopConf) {
+    return extractIcebergCatalogNames(sparkConf).stream()
+        .filter(
+            catalogName ->
+                buildHiveConf(sparkConf, hadoopConf, catalogName)
+                    .map(
+                        hiveConf -> checkDelegationTokensRequired(sparkConf, hiveConf, catalogName))
+                    .orElse(false))
+        .collect(Collectors.toSet());
+  }
+
+  private boolean checkDelegationTokensRequired(
+      SparkConf sparkConf, HiveConf hiveConf, String catalogName) {
+    String targetPrefix = SPARK_SQL_CATALOG_PREFIX + catalogName + DOT;
+
+    boolean isNeedRenewal =
+        sparkConf.getBoolean(targetPrefix + DELEGATION_TOKEN_RENEWAL_ENABLE, true);
+    boolean isHiveType = CATALOG_TYPE.equalsIgnoreCase(sparkConf.get(targetPrefix + TYPE_KEY, ""));
+    String metastoreUri = sparkConf.get(targetPrefix + URI_KEY, "");
```

Comment:
when `uri` is absent, it fallback to read from global Hadoop configuration, you should get it from `hiveConf` instead of `sparkConf` here